File: src/main/java/edu/marmara/mapper/impl/TranscriptMapperImpl.java
package edu.marmara.mapper.impl;

import edu.marmara.dto.TranscriptGetDTO;
import edu.marmara.mapper.TranscriptMapper;
import edu.marmara.model.Course;
import edu.marmara.model.Transcript;
import edu.marmara.repository.CourseRepository;
import edu.marmara.repository.impl.CourseRepositoryImpl;

import java.util.ArrayList;
import java.util.HashMap;

public class TranscriptMapperImpl implements TranscriptMapper {
    private final CourseRepository courseRepository = new CourseRepositoryImpl();

    @Override
    public Transcript mapTo(TranscriptGetDTO transcriptGetDTO) {
        Transcript transcript = new Transcript();

        transcript.setGpa(transcriptGetDTO.getGpa());
        transcript.setPassedCredit(transcriptGetDTO.getPassedCredit());
        transcript.setFailedCredit(transcriptGetDTO.getFailedCredit());

        transcript.setCurrentlyTakenCourses(new ArrayList<>());
        for (String courseCode : transcriptGetDTO.getCurrentlyTakenCourses()) {
            transcript.addCurrentlyTakenCourse(courseRepository.findByCourseCode(courseCode));
        }

        transcript.setFailedCourses(new ArrayList<>());
        for (String courseCode : transcriptGetDTO.getFailedCourses()) {
            transcript.addFailedCourse(courseRepository.findByCourseCode(courseCode));
        }

        transcript.setNotTakenCourses(new ArrayList<>());
        for (String courseCode : transcriptGetDTO.getNotTakenCourses()) {
            transcript.addNotTakenCourse(courseRepository.findByCourseCode(courseCode));
        }

        transcript.setPassedCourses(new HashMap<>());
        for (String courseAndGrade : transcriptGetDTO.getPassedCourses()) {
            String courseCode = courseAndGrade.split(" ")[0];
            Double grade = Double.valueOf(courseAndGrade.split(" ")[1]);
            transcript.getPassedCourses().put(courseRepository.findByCourseCode(courseCode), grade);
        }

        return transcript;
    }

    @Override
    public TranscriptGetDTO mapTo(Transcript transcript) {
        TranscriptGetDTO transcriptGetDTO = new TranscriptGetDTO();

        transcriptGetDTO.setGpa(transcript.getGpa());
        transcriptGetDTO.setPassedCredit(transcript.getPassedCredit());
        transcriptGetDTO.setFailedCredit(transcript.getFailedCredit());

        transcriptGetDTO.setFailedCourses(new ArrayList<>());
        for (Course course : transcript.getFailedCourses()) {
            transcriptGetDTO.addFailedCourse(course.getCourseCode());
        }

        transcriptGetDTO.setNotTakenCourses(new ArrayList<>());
        for (Course course : transcript.getNotTakenCourses()) {
            transcriptGetDTO.addNotTakenCourse(course.getCourseCode());
        }

        transcriptGetDTO.setCurrentlyTakenCourses(new ArrayList<>());
        for (Course course : transcript.getCurrentlyTakenCourses()) {
            transcriptGetDTO.addCurrentlyTakenCourse(course.getCourseCode());
        }

        transcriptGetDTO.setPassedCourses(new ArrayList<>());
        for (Course course : transcript.getPassedCourses().keySet()) {
            transcriptGetDTO.addPassesCourse(course.getCourseCode() + " " + transcript.getPassedCourses().get(course));
        }

        return transcriptGetDTO;
    }
}

File: python/src/main/marmara/mapper/impl/TranscriptMapperImpl.py


from src.main.marmara.dto.TranscriptGetDTo import TranscriptGetDTO
from src.main.marmara.mapper import TranscriptMapper
from src.main.marmara.model import Course, Transcript
from src.main.marmara.repository.impl import CourseRepositoryImpl


class TranscriptMapperImpl(TranscriptMapper):
    def __init__(self):
        self.course_repository = CourseRepositoryImpl()

    def map_to(self, transcript_get_dto: "TranscriptGetDTO") -> Transcript:
        transcript = Transcript()

        transcript.gpa = transcript_get_dto.gpa
        transcript.passed_credit = transcript_get_dto.passed_credit
        transcript.failed_credit = transcript_get_dto.failed_credit

        transcript.currently_taken_courses = []
        for course_code in transcript_get_dto.currently_taken_courses:
            transcript.currently_taken_courses.append(self.course_repository.find_by_course_code(course_code))

        transcript.failed_courses = []
        for course_code in transcript_get_dto.failed_courses:
            transcript.failed_courses.append(self.course_repository.find_by_course_code(course_code))

        transcript.not_taken_courses = []
        for course_code in transcript_get_dto.not_taken_courses:
            transcript.not_taken_courses.append(self.course_repository.find_by_course_code(course_code))

        transcript.passed_courses = {}
        for course_and_grade in transcript_get_dto.passed_courses:
            course_code, grade = course_and_grade.split(" ")
            transcript.passed_courses[self.course_repository.find_by_course_code(course_code)] = float(grade)

        return transcript

    def map_from(self, transcript: Transcript) -> "TranscriptGetDTO":
        transcript_get_dto = TranscriptGetDTO()

        transcript_get_dto.gpa = transcript.gpa
        transcript_get_dto.passed_credit = transcript.passed_credit
        transcript_get_dto.failed_credit = transcript.failed_credit

        transcript_get_dto.failed_courses = []
        for course in transcript.failed_courses:
            transcript_get_dto.failed_courses.append(course.course_code)

        transcript_get_dto.not_taken_courses = []
        for course in transcript.not_taken_courses:
            transcript_get_dto.not_taken_courses.append(course.course_code)

        transcript_get_dto.currently_taken_courses = []
        for course in transcript.currently_taken_courses:
            transcript_get_dto.currently_taken_courses.append(course.course_code)

        transcript_get_dto.passed_courses = []
        for course in transcript.passed_courses:
            transcript_get_dto.add_passed_course.append(course.course_code)

